### PR TITLE
[drop tracking] Use parent expression for scope, not parent node 

### PR DESCRIPTION
**Pull request by eholk**
Previously we were just using the parent node as the scope for a temporary value, but it turns out this is too narrow. For example, in an expression like

    Foo {
        b: &42,
        a: async { 0 }.await,
    }

the scope for the &42 was set to the ExprField node for `b: &42`, when we actually want to use the Foo struct expression.

We fix this by recursively searching through parent nodes until we find a Node::Expr. It may be that we don't find one, and if so that's okay, we will just fall back on the enclosing temporary scope which is always sufficient.

Helps with #97331

r? @jyn514

### CODE REVIEW

File: compiler/rustc_typeck/src/check/generator_interior.rs [hunk 1]
```diff
@@ -387,6 +387,18 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
             ty.needs_drop(self.fcx.tcx, self.fcx.param_env)
         };
 
+        let find_parent_expr = |mut hir_id| {
+            let hir = self.fcx.tcx.hir();
+            hir_id = hir.find_parent_node(hir_id)?;
+            loop {
+                if let hir::Node::Expr(_) = self.fcx.tcx.hir().find(hir_id)? {
```
Comment (jyn514) [inline compiler/rustc_typeck/src/check/generator_interior.rs:394]:
```suggestion
                if let hir::Node::Expr(_) = hir.find(hir_id)? {
```

File: compiler/rustc_typeck/src/check/generator_interior.rs [hunk 2]
```diff
@@ -387,6 +387,18 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
             ty.needs_drop(self.fcx.tcx, self.fcx.param_env)
         };
 
+        let find_parent_expr = |mut hir_id| {
+            let hir = self.fcx.tcx.hir();
+            hir_id = hir.find_parent_node(hir_id)?;
```
Comment (jyn514) [inline compiler/rustc_typeck/src/check/generator_interior.rs:392]: Hmm, won't this miss the case where the initial ID is already an expression? Not sure how to make a test for that.

**Review by jyn514 — COMMENTED** [review]
Are there any cases where this can have false positives? i.e. code that used to compile with drop-tracking no longer compiles?

File: compiler/rustc_typeck/src/check/generator_interior.rs [hunk 3]
```diff
@@ -387,6 +387,18 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
...
+            loop {
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/generator_interior.rs:393]: Can you use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.parent_iter here?

File: compiler/rustc_typeck/src/check/generator_interior.rs (hunk 2 above)
Comment (eholk) [reply compiler/rustc_typeck/src/check/generator_interior.rs:392]: That's intentional. We want to find the first parent node that is also an expression node. Before we just found the parent node, but now we have an additional check to see if it's an expression and if not we recur.

File: compiler/rustc_typeck/src/check/generator_interior.rs (hunk 3 above)
Comment (eholk) [reply compiler/rustc_typeck/src/check/generator_interior.rs:393]: Ah, thanks! I didn't know about that method. I think that's way cleaner, so I rewrote my code to use that. I also added a doc comment to `find_parent_node` to point people towards `parent_iter` when more appropriate.

File: compiler/rustc_typeck/src/check/generator_interior.rs (hunk 1 above)
Comment (eholk) [reply compiler/rustc_typeck/src/check/generator_interior.rs:394]: Ah, I meant to do that when I was writing this and forgot! At any rate, I completely rewrote this code using `parent_iter`.

**Comment by eholk** [comment]
> Are there any cases where this can have false positives? i.e. code that used to compile with drop-tracking no longer compiles?

I don't think so. This is only observable if an await point is in the new wider scope but wasn't in the old narrower scope. In that case, MIR would have found the type and we should have gotten an ICE saying the types didn't match. At any rate, the old behavior was incorrect (as shown by how issue-73137.rs broke with drop tracking), so it seems better to correct that.

**Pull request merged**